### ISSUE TITLE
Struggling a bit with state chart principles and Godot

### ISSUE THREAD

**Issue by eobet** [derkork/godot-statecharts#136]
I'm trying to implement a state chart with something like the following states:
* Idle
* IsMoving
* IsRotating

Now, what makes this difficult for me to grasp are two things:
* The rotation is driven by a tween which you have to await to emit a finished signal
* Godot sends different events for pressed and released inputs

So, during movement, if the player presses the rotation key and starts a rotation tween, but doesn't release the movement input key, when the rotation has finished I need to go back to the movement state _rather than the idle state_ **or** if the player has pressed the rotation key more than once this should start a new rotation (in the direction of the last rotation key pressed). I saw that history pseudo-states exist, but I can't figure out what compound states I need to set up in addition to the ones above in order to make that work (if that indeed is the preferred solution here).

In other words, I feel like my goal is to have as much of this logic as possible in State Charts, but it seems as if I must duplicate some of the state functionality with boolean flags in my code to track input actions, which somewhat defeats the purpose of the plugin (thought not really, but I hope you see my point... which is, I'm having difficulties figuring what to put in the state charts and what to put in code).

Is it possible to set this up without any duplicated state/input tracking in my code?

Sorry, is there a better forum for questions like this?

PS. I must also reveal that I'm using your plugin [in conjunction with another](https://youtu.be/LC1D0KME3gI), which eliminates the use of the expression guard for me entirely...

**Comment by smile-coder** [comment]
In my opinion, the three states of 'Idle, IsMoving, IsRotating' should actually be one state: Controllable. While in this state, the player's position and angle adjust according to the controls of the handle.

Then, you can add a state: uncontrollable.

The purpose of Statecharts is to simplify a problem, not to complicate it.

**Comment by derkork** [comment]
How about this:

![image](https://github.com/user-attachments/assets/ced802c6-b830-4245-b329-facb0b6fd5ef)


Your rotating state could have three substates which model the currently pending input. When you go into rotating, you start with the "nothing pending" substate. If the user presses rotation / movement input in this state you go to the respective "pending rotation" "pending movement" sub states. Then when the tween finishes, you transition to the pending state. I'm not sure how you want to handle multiple inputs during the tween's runtime (e.g.  player first presses rotate, then walk). This example would use the first input, but you could add a few more transitions to use the last input of the player.

**Comment by eobet** [comment]
Thank you very much for this, I think I understand the concept a little better now.

> I'm not sure how you want to handle multiple inputs during the tween's runtime (e.g. player first presses rotate, then walk). This example would use the first input, but you could add a few more transitions to use the last input of the player.

Yes, that's why I wished to be able to manipulate the historystate, because when I could flip which non-tweened movement type I should switch to after a pending tweened movement.
